### DISCUSSION TITLE
Snapshot message send order / Bank + PC

### DISCUSSION THREAD

**Discussion by verchrona** [RomanKubiak/ctrlr#639] [General]
I am programming a panel for the ML10X switcher.

I made drop down menus for the Bank 1-4, and Program 0-127.

The strange thing about this unit is that it will not change Banks unless there is a Program Change message that is behind it. So when a snapshot is sent on startup, it wont change banks unless the Bank message arrives at the unit before the PC change. 

Is there a way to order the list of messages that get sent to the hardware when snapshots are sent out?

A different solution is to program both the Bank and PC on the same dropdown menu, but I am new to CTRLR and haven't found a way to do that yet. I am assuming it is a Multi type message?

I would also love to know how to do a dropdown menu of the entire panel snapshots
![Screenshot (126)](https://github.com/RomanKubiak/ctrlr/assets/169419087/1cb22258-9931-436f-a609-f569a15ce03c)
.

**Comment by dnaldoog** [comment]
Can you post your panel?

You probably need to use _lua_ to do that, not the inbuilt  snapshot function.

**Reply by verchrona** [reply]
Here is the panel

https://file.io/4go3FVpmzyuJ

Its very simple so far, the only thing I need to sort out the the bank first, then program change. As well has snapshot menu for recall (I plan to add more knobs for a couple different attached equipment. Would be very nice to have it all change at once from this editor)

**Reply by verchrona** [reply]
I just modified it a bit, where the Multi sends both CC bank change, and Program change. But the thing is that both values go up numerically:

Preset 000: Bank 0, Program 0
Preset 001: Bank 1, Program 1

Is there a way to tell the Multi that for the first 128 programs, it stays on CC 0 Value 0 to remain on bank 1? Then from 129-256, Value 1 to go to bank 2?

https://file.io/k1phnrZzpzfV

**Comment by dnaldoog** [comment]
Does this work?

it uses _lua_:

<pre>
--
-- Called when a modulator value changes
-- @mod   http://ctrlr.org/api/class_ctrlr_modulator.html
-- @value    new numeric value of the modulator
--
sendCombinedMessage = function(--[[ CtrlrModulator --]] mod --[[ number --]], value --[[ number --]], source)
    if panel:getBootstrapState() then
        return
    end -- stop function from running onload
    local channelOut = panel:getProperty("panelMidiOutputChannelDevice")
    local statusByte = 0xB0 + (channelOut - 1) -- use this code for dynamic channel changes
    local midinumber = 0

    local bankChange = CtrlrMidiMessage({statusByte, midinumber, value})
    panel:sendMidiMessageNow(bankChange)

    -- now follow with program change

    statusByte = 0xC0 + (channelOut - 1) -- use this code for dynamic channel changes
    value = jmax(panel:getGlobalVariable(0), 0) -- if combo not set, value =-1 restrict to min 0
    --See Modulator attached to global variable in panel menu
    local programChange = CtrlrMidiMessage({statusByte, value})
    panel:sendMidiMessageNow(programChange)
end

</pre>

----
## Edit:
[Ctrlr Panel_2.zip](https://github.com/RomanKubiak/ctrlr/files/15280929/Ctrlr.Panel_2.zip)

**Reply by verchrona** [reply]
Ah wow! Yeah that was what I was looking for. One question, is there any way to make the "override panel Midi Channel" work with this controller's code? I tried checking the box, it only outputs to the selected channel in the Midi drop down menu. I will have multiple units on different channels, like the first switcher on channel 9, second one on 10.

**Reply by dnaldoog** [reply]
Do you mean this? I am getting confused 😕

<pre>
    local channelOut = panel:getProperty("panelMidiOutputChannelDevice")
    local statusByte = 0xB0 + (channelOut - 1) -- use this code for dynamic channel changes
    local midinumber = tonumber(mod:getMidiMessage():getProperty("midiMessageCtrlrNumber")) 
</pre>

**Reply by dnaldoog** [reply]
Here's a panel:

[Ctrlr Panel_3.zip](https://github.com/RomanKubiak/ctrlr/files/15281048/Ctrlr.Panel_3.zip)

**Reply by verchrona** [reply]
Sorry should have clarified. I modified this from your last upload. Basically, I duplicated everything. The top controls, I put on Channel 9. The bottom controls I put on Channel 10. The program change menus successfully are set to channels 9 and 10 and will output those correct channels on the Midi monitor. The Bank select menus are also set to channel 9 and 10, however both will continue to output on channel 4 (which is the Panel output channel) regardless if the override panel Midi Channel is selected.

https://file.io/ucdR3RP64W5I

**Reply by verchrona** [reply]
There is 1 more thing I just discovered, when I save a session in the DAW, the program change snapshot is sent on startup, but the Bank select does not.

For example, if the session is saved on Bank 1. Then I close the session, and change the bank. Then reopen the session, the Bank will not recall to Bank 1 as previously saved. Is there a way to have this sent in the session opening snapshot?

**Reply by dnaldoog** [reply]
> The Bank select menus are also set to channel 9 and 10, however both will continue to output on channel 4 (which is the Panel output channel) regardless if the override panel Midi Channel is selected.

Is this intended? It's a bit ambiguous that sentence.

**Comment by verchrona** [comment]
OK, im getting closer... I got the Bank message to stay put, while the program change is moving up. However, I am not sure how to make the bank message go to bank 2 after program change 128

https://file.io/IqWL6rsiTRLy

**Reply by dnaldoog** [reply]
>OK, im getting closer... I got the Bank message to stay put, while the program change is moving up. However, I am not sure how to make the bank message go to bank 2 after program change 128

The logic would be something like `if value > 127 then bankMess=2 end`

**Comment by dnaldoog** [comment]
> There is 1 more thing I just discovered, when I save a session in the DAW, the program change snapshot is sent on startup, but the Bank select does not.
> 
> For example, if the session is saved on Bank 1. Then I close the session, and change the bank. Then reopen the session, the Bank will not recall to Bank 1 as previously saved. Is there a way to have this sent in the session opening snapshot?

Probably remove or comment out the line(s):
<pre>
--if panel:getBootstrapState() then
--       return
--end
</pre>

**Reply by verchrona** [reply]
Thanks very much for that update. It is working on startup now. I added the knob controllers for other attached devices. I made 2 variations. One for my Guitar signal chain, one for my Bass signal chain. The units in each panel are even on different midi channels.

Triaxis > 1
EQ1 > 2
EQ2 > 3
ML10X Guitar > 4

ML10X Bass > 9
SANSAMP > 10
Analog Drive > 11
Le Bass Tube Preamp > 12 

There is one final thing I was trying to do with this panel. Since I didn't know how to create a snapshot dropdown menu, I was trying to be clever and host the panels inside Blue Cat Patchwork.. inside my main DAW.  I am using the exported restricted VST version of the Panel.

Things that work:
The panels open on startup.
The panels respond to changes made manually with the mouse.
The panels will send the correct snapshot when I click send snapshot.

Thing that doesn't work:
Blue Cat Patchwork Preset 1 (All knob settings center) changing to Blue Cat Patchwork Preset 2 (panels adjusted)

I checked the midi monitor when I click send snapshot on the panel, compared to having Blue Cat Patchwork send its own panel snapshot. When using the regular Panel (not exported restricted)... it actually works instantly and correctly because it is just like reopening the panel on startup, but keeps on opening new tab instances in each of the panels everytime a new preset is recalled as shown in the first picture.

The manual snapshot sent from inside CTRLR is correct with its midi output.
The Patchwork preset change method for some reason cuts off a ton of midi messages, and also rechannelizes all the messages to what was set in the panels (all controls with the "override panel Midi Channel" no longer override and are all sent on the same channel"

I tried different an alternative to Patchwork. Plug and Mix Chainer. What I found out from this is that instead of just changing the knob positions like Patchwork, it literally reloads the entire panel's VST. This will correctly send the right messages, but it takes about 10 seconds to reload panel, while the Patchwork is instant because it its just like changing presets on a synth instead of rebooting it.

Sending presets back and forth to panels keeps opening different tabs (after 4 Patchwork changes, a total of 5 tabs are open. It will keep doing this)
![Screenshot (129)](https://github.com/RomanKubiak/ctrlr/assets/169419087/173be728-1a0d-4f6a-9bce-7f2588dea5eb)
Default State (exported restricted VST)
![Screenshot (127)](https://github.com/RomanKubiak/ctrlr/assets/169419087/3687b6bf-40bb-49b9-808c-2801c46c011a)
1st Preset with Patchwork (exported restricted VST)
![Screenshot (128)](https://github.com/RomanKubiak/ctrlr/assets/169419087/3e9a7a42-e65e-4f31-9398-33231b9d27ca)

Is there a reason that Patchwork may not be sending the right channels and cuts off nearly 80% of the midi messages when it changes its own preset snapshots of the CTRLR panels?

Is a better solution to just figure out a way to create a snapshot dropdown menu on each? 

I attached both the standard Panels and the exported restricted VSTs of this panel.
[Guitar + Bass Panels + VST.zip](https://github.com/RomanKubiak/ctrlr/files/15284405/Guitar.%2B.Bass.Panels.%2B.VST.zip)

**Reply by dnaldoog** [reply]
It's all too much for me! I tend to be okay with actual coding algorithms (ie lua),but stuff like this does my head in. There are other people on this forum who may be able to help like Tedjuh for example, who are good with this area. I only have Cubase  - no idea what Blue Cat is although I just looked it up ;)

**Reply by verchrona** [reply]
Ok thanks for your help to bringing the panels this far into development! Hopefully others can chime in on how to figure out the jumbled midi messages from Patchwork or how to do a snapshot drop down menu.

**Reply by verchrona** [reply]
After hours of scouring for what was causing Patchwork to not recall the Panel states, I finally figured it out.

I forgot to activate "Send Snapshot on Program Change" in the Panel vst..... wow. It all works very well now when I change a Patchwork preset, it changes the settings on both the Guitar and Bass CTRLR panels at the same time.

What is a good recommended Panel snapshot delay time? I currently have it set to 10ms. Is that enough to not choke the units?

**Reply by dnaldoog** [reply]
That’s great! I have never used the snapshot send feature, so again I don’t know, but that would probably depend on the MIDI hardware, so if the panel were intended for others, maybe set up a slider to adjust that rate? My [C-TX81Z](https://sourceforge.net/projects/ctx81z/) panel has that feature.

**Reply by verchrona** [reply]
That's a cool panel. It has a lot of more complex controls than my simple one hah. Yeah, I think the 10ms is ok. I plan to make my Midi Translator Pro software double or triple all incoming messages to hardware so it makes sure all the data gets to the units and nothing gets dropped.

**Reply by dnaldoog** [reply]
Thank you! Good luck with it.
